goc/goc.go

```diff
@@ -0,0 +1,550 @@
+	case *ast.FuncDecl:
+		return handleFuncDecl(out, decl)
+	default:
+		return out.Errorf(d, "unsupported decl")
+	}
+}
+
+// handleGenDecl handles a file level declaration; a constant, a variable or an
+// import statement.
+func handleGenDecl(out *output, gd *ast.GenDecl) error {
+	for _, s := range gd.Specs {
+		switch spec := s.(type) {
+		case *ast.ValueSpec:
+			//pretty.Print(spec)
+			if err := handleValueSpec(out, spec); err != nil {
+				return err
+			}
+		case *ast.ImportSpec:
+			// Ignore imports except for comments.
+			out.Writef(s, "")
+		default:
+			return out.Errorf(s, "unsupported spec")
+		}
+		// TODO(maruel): Print spacing between declarations.
+	}
+	return nil
+}
+
+func guessType(vs *ast.ValueSpec) (token.Token, string, error) {
+	if len(vs.Values) == 0 {
+		// It is an default value, e.g. "var a int". It can't be const.
+		// token.Lookup() is not very useful as it expects "STRING" instead of
+		// "string".
+		switch n := vs.Type.(*ast.Ident).Name; n {
+		case "int":
+			return token.INT, "0", nil
+		case "string":
+			return token.STRING, "\"\"", nil
+		default:
+			return token.ILLEGAL, "", fmt.Errorf("unsupported type: %s", n)
+		}
+	}
+	// Normal declaration of type "var a = 1" or "const a = 1".
+	if len(vs.Values) > 1 {
+		return token.ILLEGAL, "", fmt.Errorf("unsupported # of values: %v", vs.Names)
+	}
+	l, ok := vs.Values[0].(*ast.BasicLit)
+	if !ok {
+		return token.ILLEGAL, "", fmt.Errorf("unsupported value: %#v", vs.Values[0])
+	}
+	return l.Kind, l.Value, nil
+}
+
+func isValueConst(vs *ast.ValueSpec) bool {
+	return vs.Names[0].Obj.Kind == ast.Con
+}
+
+// handleValueSpec handles a file level a constant or variable.
+func handleValueSpec(out *output, vs *ast.ValueSpec) error {
+	if len(vs.Names) == 0 {
+		return out.Errorf(vs, "unsupported # of value names: %v", vs.Names)
+	}
+	var decl []string
+	kind, lit, err := guessType(vs)
+	if err != nil {
+		return out.Errorf(vs, "%s", err)
+	}
+	isConst := isValueConst(vs)
+	typ := tokenStr(kind, isConst)
+	if len(typ) == 0 {
+		return out.Errorf(vs, "unsupported literal kind: %s", kind)
+	}
+	// Strictly speaking the C++ version could also define all the variables on
+	// one line but the following is easier to implement.
+	for _, name := range vs.Names {
+		out.Writef(vs, "%s;\n", strings.Join(append(decl, typ, name.Name, "=", lit), " "))
+	}
+	return nil
+}
+
+func tokenStr(kind token.Token, isConst bool) string {
+	switch kind {
+	case token.INT:
+		if isConst {
+			return "const int"
+		}
+		return "int"
+	case token.STRING:
+		if isConst {
+			return "const char * const"
+		}
+		return "const char *"
+	default:
+		return ""
+	}
+}
+
+// nodeType returns the string representation of the type of a node, in Go
+// language.
+func nodeType(n ast.Node) string {
+	if _, ok := n.(*ast.InterfaceType); ok {
+		return "interface{}"
+	}
+	if i, ok := n.(*ast.Ident); ok {
+		return i.Name
+	}
+	if _, ok := n.(*ast.FuncType); ok {
+		return "func"
```

Comment:
I'm confused, shouldn't this and `fieldToType` return the C(++) equivalent of Go types?